File: mgmt/internal/zonemd/zonemd.go
// Package zonemd computes and verifies RFC 8976 zone digests (ZONEMD, scheme SIMPLE).
package zonemd

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha512"
	"encoding/hex"
	"errors"
	"fmt"
	"hash"
	"sort"
	"strings"

	"github.com/miekg/dns"
)

// Mode is the verification setting of a secondary or RPZ transfer zone.
type Mode string

const (
	ModeOff       Mode = "off"
	ModeIfPresent Mode = "if_present"
	ModeRequired  Mode = "required"
)

// Status is the outcome of a verification.
type Status string

const (
	StatusOff      Status = "off"
	StatusAbsent   Status = "absent"
	StatusVerified Status = "verified"
	StatusFailed   Status = "failed"
)

// Result is a verification outcome; Err names the reason when Status is failed.
type Result struct {
	Status Status
	Err    string
}

// ZONEMD scheme and hash algorithm numbers (RFC 8976 §5).
const (
	SchemeSimple uint8 = 1
	HashSHA384   uint8 = 1
	HashSHA512   uint8 = 2
)

// Digest returns the SIMPLE-scheme digest of the zone at origin over rrs with the given hash.
func Digest(origin string, rrs []dns.RR, hashAlg uint8) ([]byte, error) {
	var h hash.Hash
	switch hashAlg {
	case HashSHA384:
		h = sha512.New384()
	case HashSHA512:
		h = sha512.New()
	default:
		return nil, fmt.Errorf("unsupported ZONEMD hash %d", hashAlg)
	}
	wire, err := canonical(origin, rrs)
	if err != nil {
		return nil, err
	}
	for _, w := range wire {
		h.Write(w.packed)
	}
	return h.Sum(nil), nil
}

// Verify checks the apex ZONEMD records of the zone at origin against rrs (RFC 8976 §4).
func Verify(origin string, rrs []dns.RR, mode Mode) Result {
	if mode != ModeIfPresent && mode != ModeRequired {
		return Result{Status: StatusOff}
	}
	origin = dns.CanonicalName(origin)
	var soa *dns.SOA
	var apex []*dns.ZONEMD
	for _, rr := range rrs {
		if dns.CanonicalName(rr.Header().Name) != origin {
			continue
		}
		switch r := rr.(type) {
		case *dns.SOA:
			soa = r
		case *dns.ZONEMD:
			apex = append(apex, r)
		}
	}
	if len(apex) == 0 {
		if mode == ModeRequired {
			return Result{Status: StatusFailed, Err: "no apex ZONEMD"}
		}
		return Result{Status: StatusAbsent}
	}
	if soa == nil {
		return Result{Status: StatusFailed, Err: "no apex SOA"}
	}
	type tuple struct{ scheme, hash uint8 }
	seen := map[tuple]int{}
	for _, z := range apex {
		seen[tuple{z.Scheme, z.Hash}]++
	}
	reason := "no supported ZONEMD"
	digests := map[uint8][]byte{}
	for _, z := range apex {
		if seen[tuple{z.Scheme, z.Hash}] > 1 {
			reason = "duplicate scheme and hash"
			continue
		}
		if z.Serial != soa.Serial {
			reason = "serial mismatch"
			continue
		}
		if z.Scheme != SchemeSimple || (z.Hash != HashSHA384 && z.Hash != HashSHA512) {
			continue
		}
		want, err := hex.DecodeString(z.Digest)
		if err != nil || len(want) < 12 || (z.Hash == HashSHA384 && len(want) != 48) || (z.Hash == HashSHA512 && len(want) != 64) {
			reason = "digest mismatch"
			continue
		}
		got, ok := digests[z.Hash]
		if !ok {
			if got, err = Digest(origin, rrs, z.Hash); err != nil {
				reason = err.Error()
				continue
			}
			digests[z.Hash] = got
		}
		if hmac.Equal(got, want) {
			return Result{Status: StatusVerified}
		}
		reason = "digest mismatch"
	}
	return Result{Status: StatusFailed, Err: reason}
}

// Placeholder returns an apex ZONEMD (SIMPLE, SHA-384) with an all-zero digest, to be filled by Apply.
func Placeholder(origin string, serial, ttl uint32) *dns.ZONEMD {
	return &dns.ZONEMD{
		Hdr:    dns.RR_Header{Name: dns.CanonicalName(origin), Rrtype: dns.TypeZONEMD, Class: dns.ClassINET, Ttl: ttl},
		Serial: serial,
		Scheme: SchemeSimple,
		Hash:   HashSHA384,
		Digest: strings.Repeat("0", 96),
	}
}

// Apply sets the serial (from the apex SOA) and the SHA-384 digest of the apex ZONEMD in rrs, in place.
func Apply(origin string, rrs []dns.RR) error {
	canon := dns.CanonicalName(origin)
	var soa *dns.SOA
	var z *dns.ZONEMD
	for _, rr := range rrs {
		if dns.CanonicalName(rr.Header().Name) != canon {
			continue
		}
		switch r := rr.(type) {
		case *dns.SOA:
			soa = r
		case *dns.ZONEMD:
			if z == nil {
				z = r
			}
		}
	}
	if z == nil {
		return errors.New("zonemd: no apex ZONEMD")
	}
	if soa == nil {
		return errors.New("zonemd: no apex SOA")
	}
	z.Serial, z.Scheme, z.Hash = soa.Serial, SchemeSimple, HashSHA384
	d, err := Digest(origin, rrs, HashSHA384)
	if err != nil {
		return err
	}
	z.Digest = hex.EncodeToString(d)
	return nil
}

type canonRR struct {
	packed []byte
	labels [][]byte // owner labels, root first
	rrtype uint16
	class  uint16
	rdata  []byte
}

// canonical returns the zone's RRs in RFC 8976 §3.3 canonical wire form and order: owners at or
// below origin, apex ZONEMD and apex RRSIG(ZONEMD) excluded, owner and RFC 4034 §6.2 RDATA names
// lowercased, sorted by owner, type and RDATA, duplicates once.
func canonical(origin string, rrs []dns.RR) ([]canonRR, error) {
	origin = dns.CanonicalName(origin)
	out := make([]canonRR, 0, len(rrs))
	buf := make([]byte, 65535+512)
	for _, rr := range rrs {
		owner := dns.CanonicalName(rr.Header().Name)
		if !dns.IsSubDomain(origin, owner) {
			continue
		}
		if owner == origin {
			if rr.Header().Rrtype == dns.TypeZONEMD {
				continue
			}
			if s, ok := rr.(*dns.RRSIG); ok && s.TypeCovered == dns.TypeZONEMD {
				continue
			}
		}
		c := dns.Copy(rr)
		c.Header().Name = owner
		lowerRdataNames(c)
		n, err := dns.PackRR(c, buf, 0, nil, false)
		if err != nil {
			return nil, fmt.Errorf("zonemd: pack %s: %w", c.Header().Name, err)
		}
		packed := append([]byte(nil), buf[:n]...)
		ownerLen, err := wireNameLen(packed)
		if err != nil || ownerLen+10 > n {
			return nil, fmt.Errorf("zonemd: malformed wire form of %s", c.Header().Name)
		}
		// Lowercase the owner octets too, which also covers escaped letters (\065); label length
		// octets are at most 63 and never in 'A'..'Z'.
		for i := 0; i < ownerLen; i++ {
			if packed[i] >= 'A' && packed[i] <= 'Z' {
				packed[i] += 'a' - 'A'
			}
		}
		out = append(out, canonRR{
			packed: packed,
			labels: ownerLabels(packed[:ownerLen]),
			rrtype: c.Header().Rrtype,
			class:  c.Header().Class,
			rdata:  packed[ownerLen+10:],
		})
	}
	sort.SliceStable(out, func(i, j int) bool {
		a, b := out[i], out[j]
		if k := compareOwners(a.labels, b.labels); k != 0 {
			return k < 0
		}
		if a.rrtype != b.rrtype {
			return a.rrtype < b.rrtype
		}
		if a.class != b.class {
			return a.class < b.class
		}
		return bytes.Compare(a.rdata, b.rdata) < 0
	})
	uniq := out[:0]
	for i, r := range out {
		if i > 0 {
			p := uniq[len(uniq)-1]
			if compareOwners(p.labels, r.labels) == 0 && p.rrtype == r.rrtype && p.class == r.class && bytes.Equal(p.rdata, r.rdata) {
				continue
			}
		}
		uniq = append(uniq, r)
	}
	return uniq, nil
}

// wireNameLen returns the length of the uncompressed wire name at the start of b.
func wireNameLen(b []byte) (int, error) {
	i := 0
	for i < len(b) {
		l := int(b[i])
		if l == 0 {
			return i + 1, nil
		}
		if l > 63 {
			return 0, errors.New("compressed or invalid label")
		}
		i += 1 + l
	}
	return 0, errors.New("truncated name")
}

// ownerLabels returns the labels of an uncompressed wire name, root first (RFC 4034 §6.1 order).
func ownerLabels(wire []byte) [][]byte {
	var labels [][]byte
	for i := 0; wire[i] != 0; i += 1 + int(wire[i]) {
		labels = append(labels, wire[i+1:i+1+int(wire[i])])
	}
	for i, j := 0, len(labels)-1; i < j; i, j = i+1, j-1 {
		labels[i], labels[j] = labels[j], labels[i]
	}
	return labels
}

// compareOwners orders names by RFC 4034 §6.1: labels compared from the root as unsigned octets, a
// name sorting before the names below it.
func compareOwners(a, b [][]byte) int {
	for i := 0; i < len(a) && i < len(b); i++ {
		if c := bytes.Compare(a[i], b[i]); c != 0 {
			return c
		}
	}
	return len(a) - len(b)
}

// lowerRdataNames lowercases the RDATA domain names of the RFC 4034 §6.2 types (as corrected by
// RFC 6840 §5.1: not NSEC, not HINFO).
func lowerRdataNames(rr dns.RR) {
	l := strings.ToLower
	switch r := rr.(type) {
	case *dns.NS:
		r.Ns = l(r.Ns)
	case *dns.MD:
		r.Md = l(r.Md)
	case *dns.MF:
		r.Mf = l(r.Mf)
	case *dns.CNAME:
		r.Target = l(r.Target)
	case *dns.SOA:
		r.Ns, r.Mbox = l(r.Ns), l(r.Mbox)
	case *dns.MB:
		r.Mb = l(r.Mb)
	case *dns.MG:
		r.Mg = l(r.Mg)
	case *dns.MR:
		r.Mr = l(r.Mr)
	case *dns.PTR:
		r.Ptr = l(r.Ptr)
	case *dns.MINFO:
		r.Rmail, r.Email = l(r.Rmail), l(r.Email)
	case *dns.MX:
		r.Mx = l(r.Mx)
	case *dns.RP:
		r.Mbox, r.Txt = l(r.Mbox), l(r.Txt)
	case *dns.AFSDB:
		r.Hostname = l(r.Hostname)
	case *dns.RT:
		r.Host = l(r.Host)
	case *dns.SIG:
		r.SignerName = l(r.SignerName)
	case *dns.RRSIG:
		r.SignerName = l(r.SignerName)
	case *dns.PX:
		r.Map822, r.Mapx400 = l(r.Map822), l(r.Mapx400)
	case *dns.NXT:
		r.NextDomain = l(r.NextDomain)
	case *dns.NAPTR:
		r.Replacement = l(r.Replacement)
	case *dns.KX:
		r.Exchanger = l(r.Exchanger)
	case *dns.SRV:
		r.Target = l(r.Target)
	case *dns.DNAME:
		r.Target = l(r.Target)
	}
}

File: engine/src/zonemd.rs
//! RFC 8976 zone digests (ZONEMD): digest and verify.
//!
//! The canonical form follows RFC 8976 §3.3: owners lowercased, RDATA names lowercased for the
//! RFC 4034 §6.2 types (as corrected by RFC 6840 §5.1), RRs sorted in RFC 4034 §6.1 order, then by
//! type and RDATA octets, duplicates included once, apex ZONEMD and its RRSIGs excluded. The
//! management plane (`mgmt/internal/zonemd`) implements the same digest; both are pinned to the
//! RFC 8976 Appendix A vectors. Verification runs on the control runtime, never on the query path.

use hickory_proto::dnssec::rdata::DNSSECRData;
use hickory_proto::rr::{Name, RData, Record};
use hickory_proto::serialize::binary::{BinEncodable, BinEncoder, NameEncoding};
use sha2::{Digest as _, Sha384, Sha512};

use crate::proto::{ZonemdStatus, ZonemdVerify};

pub const TYPE_ZONEMD: u16 = 63;
pub const SCHEME_SIMPLE: u8 = 1;
pub const HASH_SHA384: u8 = 1;
pub const HASH_SHA512: u8 = 2;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum VerifyMode {
    Off,
    IfPresent,
    Required,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Verdict {
    Off,
    Absent,
    Verified,
    Failed(String),
}

impl VerifyMode {
    /// `ZONEMD_VERIFY_UNSPECIFIED` (an older management plane) and unknown values mean off.
    pub fn from_proto(v: i32) -> VerifyMode {
        if v == ZonemdVerify::IfPresent as i32 {
            VerifyMode::IfPresent
        } else if v == ZonemdVerify::Required as i32 {
            VerifyMode::Required
        } else {
            VerifyMode::Off
        }
    }
}

impl Verdict {
    /// The `ZonemdStatus` value and the error text (empty unless failed).
    pub fn to_proto(&self) -> (i32, String) {
        match self {
            Verdict::Off => (ZonemdStatus::Off as i32, String::new()),
            Verdict::Absent => (ZonemdStatus::Absent as i32, String::new()),
            Verdict::Verified => (ZonemdStatus::Verified as i32, String::new()),
            Verdict::Failed(e) => (ZonemdStatus::Failed as i32, e.clone()),
        }
    }
}

/// One RR in canonical wire form with its sort key.
struct Canonical {
    /// Lowercase owner labels, root side first: `Vec` ordering is RFC 4034 §6.1 ordering.
    labels: Vec<Vec<u8>>,
    rtype: u16,
    wire: Vec<u8>,
    /// Offset of the RDATA in `wire`.
    rdata: usize,
}

fn canonical(origin: &Name, records: &[Record]) -> Result<Vec<Canonical>, String> {
    let mut out = Vec::with_capacity(records.len());
    for r in records {
        if !origin.zone_of(&r.name) {
            continue;
        }
        let rtype = u16::from(r.record_type());
        let covers_zonemd = matches!(&r.data, RData::DNSSEC(DNSSECRData::RRSIG(s)) if u16::from(s.input().type_covered) == TYPE_ZONEMD);
        if r.name == *origin && (rtype == TYPE_ZONEMD || covers_zonemd) {
            continue;
        }
        let mut wire = Vec::new();
        let mut enc = BinEncoder::new(&mut wire);
        // Canonical form lowercases RDATA names of the types hickory decodes (RFC 4034 §6.2).
        enc.set_canonical_form(true);
        enc.set_name_encoding(NameEncoding::UncompressedLowercase);
        r.emit(&mut enc)
            .map_err(|e| format!("encode {} {}: {e}", r.name, r.record_type()))?;
        let rdata = r.name.iter().map(|l| l.len() + 1).sum::<usize>() + 1 + 10;
        if rdata > wire.len() {
            return Err(format!(
                "encode {} {}: short record",
                r.name,
                r.record_type()
            ));
        }
        if let RData::Unknown { .. } = r.data {
            lowercase_opaque_names(rtype, &mut wire[rdata..]);
        }
        let labels = r
            .name
            .iter()
            .rev()
            .map(|l| l.to_ascii_lowercase())
            .collect();
        out.push(Canonical {
            labels,
            rtype,
            wire,
            rdata,
        });
    }
    out.sort_by(|a, b| {
        a.labels
            .cmp(&b.labels)
            .then(a.rtype.cmp(&b.rtype))
            .then_with(|| a.wire[a.rdata..].cmp(&b.wire[b.rdata..]))
    });
    // Equal owner, type, class and RDATA are digested once (the TTL does not distinguish them).
    out.dedup_by(|b, a| {
        a.wire[..a.rdata - 6] == b.wire[..b.rdata - 6] && a.wire[a.rdata..] == b.wire[b.rdata..]
    });
    Ok(out)
}

/// Lowercases the RDATA names of RFC 4034 §6.2 types that hickory keeps as opaque bytes.
/// Names carrying compression pointers are left unchanged (RFC 3597 §4 forbids them there).
fn lowercase_opaque_names(rtype: u16, rd: &mut [u8]) {
    let (skip, names) = match rtype {
        3 | 4 | 7 | 8 | 9 | 30 | 39 => (0, 1), // MD MF MB MG MR NXT DNAME
        14 | 17 => (0, 2),                     // MINFO RP
        18 | 21 | 36 => (2, 1),                // AFSDB RT KX
        26 => (2, 2),                          // PX
        38 => match rd.first() {
            // A6: prefix length, address suffix octets, prefix name when the prefix is not empty
            Some(&p) if p > 0 && p <= 128 => (1 + (128 - usize::from(p)).div_ceil(8), 1),
            _ => return,
        },
        _ => return,
    };
    let mut pos = skip;
    for _ in 0..names {
        loop {
            let Some(&len) = rd.get(pos) else { return };
            if len == 0 {
                pos += 1;
                break;
            }
            if len & 0xc0 != 0 {
                return;
            }
            let end = pos + 1 + usize::from(len);
            let Some(label) = rd.get_mut(pos + 1..end) else {
                return;
            };
            label.make_ascii_lowercase();
            pos = end;
        }
    }
}

/// The RFC 8976 SIMPLE-scheme digest of the zone at `origin` with SHA-384 or SHA-512.
pub fn digest(origin: &Name, records: &[Record], hash: u8) -> Result<Vec<u8>, String> {
    let rrs = canonical(origin, records)?;
    hash_canonical(&rrs, hash)
}

fn hash_canonical(rrs: &[Canonical], hash: u8) -> Result<Vec<u8>, String> {
    match hash {
        HASH_SHA384 => {
            let mut h = Sha384::new();
            rrs.iter().for_each(|c| h.update(&c.wire));
            Ok(h.finalize().to_vec())
        }
        HASH_SHA512 => {
            let mut h = Sha512::new();
            rrs.iter().for_each(|c| h.update(&c.wire));
            Ok(h.finalize().to_vec())
        }
        _ => Err(format!("unsupported hash algorithm {hash}")),
    }
}

/// Verifies the apex ZONEMD RRs of `records` (RFC 8976 §4). The first usable RR whose digest
/// matches verifies the zone; otherwise the verdict names the last reason.
pub fn verify(origin: &Name, records: &[Record], mode: VerifyMode) -> Verdict {
    if mode == VerifyMode::Off {
        return Verdict::Off;
    }
    let zonemds: Vec<&[u8]> = records
        .iter()
        .filter(|r| r.name == *origin && u16::from(r.record_type()) == TYPE_ZONEMD)
        .map(|r| match &r.data {
            RData::Unknown { rdata, .. } => rdata.anything.as_slice(),
            _ => &[],
        })
        .collect();
    if zonemds.is_empty() {
        return match mode {
            VerifyMode::IfPresent => Verdict::Absent,
            _ => Verdict::Failed("no apex ZONEMD".into()),
        };
    }
    let soa_serial = records.iter().find_map(|r| match &r.data {
        RData::SOA(soa) if r.name == *origin => Some(soa.serial),
        _ => None,
    });
    let mut rrs: Option<Vec<Canonical>> = None;
    let mut reason = String::from("no supported ZONEMD");
    for rd in &zonemds {
        if rd.len() < 6 {
            reason = "no supported ZONEMD".into();
            continue;
        }
        if zonemds
            .iter()
            .filter(|o| o.len() >= 6 && o[4..6] == rd[4..6])
            .count()
            > 1
        {
            reason = "duplicate scheme and hash".into();
            continue;
        }
        if soa_serial != Some(u32::from_be_bytes([rd[0], rd[1], rd[2], rd[3]])) {
            reason = "serial mismatch".into();
            continue;
        }
        let want_len = match (rd[4], rd[5]) {
            (SCHEME_SIMPLE, HASH_SHA384) => 48,
            (SCHEME_SIMPLE, HASH_SHA512) => 64,
            _ => {
                reason = "no supported ZONEMD".into();
                continue;
            }
        };
        if rd.len() - 6 != want_len {
            reason = "digest length mismatch".into();
            continue;
        }
        if rrs.is_none() {
            match canonical(origin, records) {
                Ok(c) => rrs = Some(c),
                Err(e) => return Verdict::Failed(e),
            }
        }
        match hash_canonical(rrs.as_deref().unwrap_or_default(), rd[5]) {
            Ok(d) if d == rd[6..] => return Verdict::Verified,
            Ok(_) => reason = "digest mismatch".into(),
            Err(e) => reason = e,
        }
    }
    Verdict::Failed(reason)
}

#[cfg(test)]
mod tests {
    use super::*;
    use hickory_proto::rr::{Name, RData, Record, RecordType};
    use hickory_proto::serialize::binary::BinDecodable;

    fn vector(name: &str) -> Vec<Record> {
        let path = format!(
            "{}/../e2e/testdata/rfc8976/{name}.hex",
            env!("CARGO_MANIFEST_DIR")
        );
        std::fs::read_to_string(&path)
            .unwrap()
            .lines()
            .map(|l| Record::from_bytes(&hex_decode(l)).unwrap())
            .collect()
    }

    fn hex_decode(s: &str) -> Vec<u8> {
        (0..s.len())
            .step_by(2)
            .map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap())
            .collect()
    }

    fn zonemd_rdata(r: &Record) -> Option<&[u8]> {
        match &r.data {
            RData::Unknown { rdata, .. } if u16::from(r.record_type()) == TYPE_ZONEMD => {
                Some(&rdata.anything)
            }
            _ => None,
        }
    }

    #[test]
    fn digest_matches_rfc8976_appendix_a() {
        let origin = Name::from_ascii("example.").unwrap();
        let mut checked = 0;
        for name in ["a1", "a2", "a3"] {
            let rrs = vector(name);
            for r in &rrs {
                let Some(rd) = zonemd_rdata(r) else { continue };
                if r.name != origin
                    || rd[4] != SCHEME_SIMPLE
                    || !(rd[5] == HASH_SHA384 || rd[5] == HASH_SHA512)
                {
                    continue;
                }
                assert_eq!(
                    digest(&origin, &rrs, rd[5]).unwrap(),
                    rd[6..].to_vec(),
                    "{name} hash {}",
                    rd[5]
                );
                checked += 1;
            }
            assert_eq!(
                verify(&origin, &rrs, VerifyMode::Required),
                Verdict::Verified,
                "{name}"
            );
        }
        assert_eq!(checked, 4);
    }

    #[test]
    fn verify_rules() {
        let origin = Name::from_ascii("example.").unwrap();
        let base = vector("a1");
        let is_apex_zonemd =
            |r: &Record| u16::from(r.record_type()) == TYPE_ZONEMD && r.name == origin;
        let with_apex = |f: &dyn Fn(&mut Vec<u8>)| -> Vec<Record> {
            base.iter()
                .map(|r| {
                    if !is_apex_zonemd(r) {
                        return r.clone();
                    }
                    let mut rd = zonemd_rdata(r).unwrap().to_vec();
                    f(&mut rd);
                    Record::from_rdata(
                        r.name.clone(),
                        r.ttl,
                        RData::Unknown {
                            code: RecordType::Unknown(TYPE_ZONEMD),
                            rdata: hickory_proto::rr::rdata::NULL::with(rd),
                        },
                    )
                })
                .collect()
        };
        let without: Vec<Record> = base
            .iter()
            .filter(|r| !is_apex_zonemd(r))
            .cloned()
            .collect();
        assert_eq!(
            verify(&origin, &base, VerifyMode::IfPresent),
            Verdict::Verified
        );
        assert_eq!(
            verify(&origin, &with_apex(&|rd| rd[6] ^= 1), VerifyMode::Off),
            Verdict::Off
        );
        assert!(
            matches!(verify(&origin, &with_apex(&|rd| rd[3] ^= 1), VerifyMode::IfPresent), Verdict::Failed(e) if e.contains("serial"))
        );
        assert!(
            matches!(verify(&origin, &with_apex(&|rd| rd[6] ^= 1), VerifyMode::IfPresent), Verdict::Failed(e) if e.contains("digest"))
        );
        assert!(matches!(
            verify(
                &origin,
                &with_apex(&|rd| rd[5] = 200),
                VerifyMode::IfPresent
            ),
            Verdict::Failed(_)
        ));
        assert!(matches!(
            verify(
                &origin,
                &with_apex(&|rd| rd.truncate(6 + 11)),
                VerifyMode::IfPresent
            ),
            Verdict::Failed(_)
        ));
        let mut dup = base.clone();
        dup.extend(
            with_apex(&|rd| rd[7] ^= 1)
                .into_iter()
                .filter(|r| is_apex_zonemd(r)),
        );
        assert!(
            matches!(verify(&origin, &dup, VerifyMode::IfPresent), Verdict::Failed(e) if e.contains("duplicate"))
        );
        assert_eq!(
            verify(&origin, &without, VerifyMode::IfPresent),
            Verdict::Absent
        );
        assert!(matches!(
            verify(&origin, &without, VerifyMode::Required),
            Verdict::Failed(_)
        ));
        assert_eq!(VerifyMode::from_proto(0), VerifyMode::Off);
        assert_eq!(Verdict::Verified.to_proto().0, 3);
    }

    // Catches: RDATA names of RFC 4034 §6.2 types hickory keeps opaque (DNAME, RP, AFSDB) digested
    // with their zone-file case, so the engine and the management plane disagree on such zones.
    #[test]
    fn opaque_rdata_names_are_lowercased() {
        let origin = Name::from_ascii("example.").unwrap();
        let zone = |dname: &[u8], rp: &[u8], afsdb: &[u8], txt: &str| -> Vec<Record> {
            let opaque = |owner: &str, code: u16, rd: &[u8]| {
                Record::from_rdata(
                    Name::from_ascii(owner).unwrap(),
                    300,
                    RData::Unknown {
                        code: RecordType::Unknown(code),
                        rdata: hickory_proto::rr::rdata::NULL::with(rd.to_vec()),
                    },
                )
            };
            vec![
                opaque("d.example.", 39, dname),
                opaque("r.example.", 17, rp),
                opaque("a.example.", 18, afsdb),
                Record::from_rdata(
                    Name::from_ascii("t.example.").unwrap(),
                    300,
                    RData::TXT(hickory_proto::rr::rdata::TXT::new(vec![txt.into()])),
                ),
            ]
        };
        let lower = zone(
            b"\x01x\x07example\x00",
            b"\x01m\x00\x01t\x00",
            b"\x00\x01\x01h\x00",
            "Case",
        );
        let upper = zone(
            b"\x01X\x07EXAMPLE\x00",
            b"\x01M\x00\x01T\x00",
            b"\x00\x01\x01H\x00",
            "Case",
        );
        let text_differs = zone(
            b"\x01x\x07example\x00",
            b"\x01m\x00\x01t\x00",
            b"\x00\x01\x01h\x00",
            "case",
        );
        let d = digest(&origin, &lower, HASH_SHA384).unwrap();
        assert_ne!(
            d,
            digest(&origin, &text_differs, HASH_SHA384).unwrap(),
            "TXT case is data"
        );
        assert_eq!(d, digest(&origin, &upper, HASH_SHA384).unwrap());
    }
}
